taigaio/taiga
# feat(events): i#2853 unsubscribe from projetc events if project permissions changes

![](https://media.giphy.com/media/njYrp176NQsHS/giphy-downsized-large.gif)

This PR resolves error i#2853. Create a system to close all the project events subscriptions related to those users who lose their privileges.

There are many events that can cause this loss (loss of permissions for users external to the project or workspace, loss of project or workspace membership, user deletion…) For this reason, the system is explicit. A new event will be sent from the service, when a situation that could be generated a loss of privilege, so that all the subscriptions of the associated project will be checked.

The Subscriber object will detect those EventResponses that contain actions to execute instead of sending to clients, that's all. Therefore, there is a new action to unsubscribe if the user has lost privileges to the project and a function to emit the event from the service.

Currently only the change of permissions for users who are not members of the project is implemented.

This PR adds unit tests for all the Subscriber class too.

## How to validate:
- [x] Test are green
  - You need one project and three users (the superadmin panel can be useful to design this scenario)
    - `admin`: the project admin
    - `user1`: is a member of the workspace but is not a member of the project
    - `user2`: is not a member of the project or the workspace.
  - [x] Test for workspace member users.
    - With `admin`, open `Project Settings > Roles and Permissions > Workspace permissions` to set `can view` 
    - With `user1`, go to the project kanban, open the dev tools and check the events connection is working properly.
    - With `admin` change `Workspace permissions` to `no access`
    - With `user1`, check events received, there is one response with type `action`, command `check_project_events_subscription`, status `error` and detail `lost-permissions`.
  - [x] Test for not project or workspace member users.
    - With `admin`, open `Project Settings > Roles and Permissions > Public permissions` to set `can view` 
    - With `user2`, go to the project kanban, open the dev tools and check the events connection is working properly.
    - With `admin` change `Public permissions` to `no access`
    - With `user2`, check events received, there is one response with type `action`, command `check_project_events_subscription`, status `error` and detail `lost-permissions`.
